nixos/modules/services/web-apps/weblate.nix

```diff
@@ -0,0 +1,421 @@
+      after = [
+        "network.target"
+        "redis.service"
+        "postgresql.service"
+      ];
+      environment = environment // {
+        CELERY_WORKER_RUNNING = "1";
+      };
+      path = weblatePath;
+      # Recommendations from:
+      # https://github.com/WeblateOrg/weblate/blob/main/weblate/examples/celery-weblate.service
+      serviceConfig =
+        let
+          # We have to push %n through systemd's replacement, therefore %%n.
+          pidFile = "/run/celery/weblate-%%n.pid";
+          nodes = "celery notify memory backup translate";
+          cmd = verb: ''
+            ${pythonEnv}/bin/celery multi ${verb} \
+              ${nodes} \
+              -A "weblate.utils" \
+              --pidfile=${pidFile} \
+              --logfile=/var/log/celery/weblate-%%n%%I.log \
+              --loglevel=DEBUG \
+              --beat:celery \
+              --queues:celery=celery \
+              --prefetch-multiplier:celery=4 \
+              --queues:notify=notify \
+              --prefetch-multiplier:notify=10 \
+              --queues:memory=memory \
+              --prefetch-multiplier:memory=10 \
+              --queues:translate=translate \
+              --prefetch-multiplier:translate=4 \
+              --concurrency:backup=1 \
+              --queues:backup=backup \
+              --prefetch-multiplier:backup=2
+          '';
+        in
+        {
+          Type = "forking";
+          User = "weblate";
+          Group = "weblate";
+          WorkingDirectory = "${finalPackage}/${python.sitePackages}/weblate/";
+          RuntimeDirectory = "celery";
+          RuntimeDirectoryPreserve = "restart";
+          LogsDirectory = "celery";
+          ExecStart = cmd "start";
+          ExecReload = cmd "restart";
+          ExecStop = ''
+            ${finalPackage}/bin/celery multi stopwait \
+              ${nodes} \
+              --pidfile=${pidFile}
+          '';
+          Restart = "always";
+        };
+    };
+
+    systemd.services.weblate = {
+      description = "Weblate uWSGI app";
+      after = [
+        "network.target"
+        "postgresql.service"
+        "redis.service"
+        "weblate-migrate.service"
+        "weblate-postgresql-setup.service"
+      ];
+      requires = [
+        "weblate-migrate.service"
+        "weblate-postgresql-setup.service"
+        "weblate-celery.service"
+        "weblate.socket"
+      ];
+      inherit environment;
+      path = weblatePath;
+      serviceConfig = {
+        Type = "notify";
+        NotifyAccess = "all";
+        ExecStart =
+          let
+            uwsgi = pkgs.uwsgi.override {
+              plugins = [ "python3" ];
+              python3 = python;
+            };
+            jsonConfig = pkgs.writeText "uwsgi.json" (builtins.toJSON uwsgiConfig);
+          in
+          "${uwsgi}/bin/uwsgi --json ${jsonConfig}";
+        Restart = "on-failure";
+        KillSignal = "SIGTERM";
+        WorkingDirectory = finalPackage;
```

Comment:
[Upstreams docs](https://docs.weblate.org/en/latest/admin/install.html#uwsgi) seem to tie uwsgi to Ngnix and gunicorn to Apache, even though it obviously seems to run differently. For now I think I'd rather follow upstreams setup recommendations with Nginx/uwsgi as closely, to be confident in the configuration.

I had my struggles with uwsgi, but didn't notice any problems recently.

Reply:
I'm reconsidering using unicorn. Could you share your full code with me? I think the snippet doesn't contain how you expose the `uwsgi` module to gunicorn.

Reply:
I think that comes from `weblate/wsgi.py` in the weblate repository. You just have to set the `WorkingDirectory` appropriately. Not sure if that's still `dataDir` or if it would be `${finalPackage}/${python.sitePackages}/weblate/`

Reply:
That worked, thanks!